liquidity/threshold_rule.go

```diff
@@ -0,0 +1,229 @@
+	)
+
+	return newSwapSet(reason, swaps), nil
+}
+
+// shouldSwap examines our current set of balances, and required thresholds and
+// determines whether we can improve our liquidity balance. It returns a swap
+// type which indicates the type of swap that should be executed (and nil if
+// no swap is possible/required) and a reason enum which further explain the
+// reasoning for this action.
+func shouldSwap(balances *balances, minInbound, minOutbound int) (*swap.Type,
+	Reason) {
+
+	// If we have no channels open (which no capacity indicates), we cannot
+	// recommend any loops.
+	if balances.capacity == 0 {
+		return nil, ReasonNoCapacity
+	}
+
+	// Calculate the current percentage of inbound and outbound capacity
+	// that we have.
+	currentInbound := balances.incoming * 100 / balances.capacity
+	currentOutbound := balances.outgoing * 100 / balances.capacity
+
+	inboundDeficit := uint64(currentInbound) < uint64(minInbound)
+	outboundDeficit := uint64(currentOutbound) < uint64(minOutbound)
+
+	switch {
+	// If we have too little inbound and too little outbound, we cannot
+	// do anything to help our liquidity situation. This will happen in the
+	// case where we have a lot of pending htlcs on our channels.
+	case inboundDeficit && outboundDeficit:
+		return nil, ReasonNoSurplus
+
+	// If we have too little inbound, but not too little outbound, it is
+	// possible that a loop out will improve our inbound liquidity.
+	case inboundDeficit && !outboundDeficit:
+		out := swap.TypeOut
+		return &out, ReasonImbalanced
+
+	// If we have enough inbound, and too little outbound, it is possible
+	// that we can loop in to improve our outbound liquidity.
+	case !inboundDeficit && outboundDeficit:
+		in := swap.TypeIn
+		return &in, ReasonImbalanced
+
+	// If we have enough inbound and enough outbound, we do not need to
+	// take any actions at present.
+	default:
+		return nil, ReasonLiquidityOk
+	}
+}
+
+// calculateSwapAmount calculates the amount of capacity that we need to shift
+// to improve the balance of our channels, and returns a reason which provides
+// additional context. This function is used in the case where we have surplus
+// liquidity in one direction, and deficit in another. This is the case in which
+// we recommend swaps, if we have deficits on both sides, we cannot swap without
+// further unbalancing, and if we have surplus on both sides, we do not need to.
+//
+// This function calculates the portion of our total capacity we should shift
+// from the surplus side to deficit side without unbalancing the surplus side.
+// This is important, because we do not want to recommend a swap in one
+// direction that just results in our needing to produce a swap in the other
+// direction.
+//
+// If we do have enough surplus, we aim to fall in the midpoint of our two
+// minimums so that we do not tip our channel balance too far in one direction
+// or the other.
+func calculateSwapAmount(capacity, deficitCurrent, deficitMinimum,
```

Comment:
`deficitMinimum` as a name confused me a bit. I read it as the 'minimally required deficit', while it really is 'minimum balance on the deficit side'?.